lib/room.js

```diff
@@ -360,11 +387,19 @@ function handleSignalingEvents(room, signaling) {
 
   // Reemit state transition events from the RoomSignaling.
   signaling.on('stateChanged', function stateChanged(state, error) {
-    room.participants.forEach(participant => {
-      participant._unsubscribeTracks();
-    });
     log.info('Transitioned to state:', state);
-    room.emit(state, room, error);
+    if (state === 'disconnected') {
+      room.participants.forEach(participant => {
+        participant._unsubscribeTracks();
+      });
+      room.emit(state, room, error);
+      signaling.removeListener('stateChanged', stateChanged);
+      return;
+    } else if (state === 'reconnecting') {
```

Comment:
@markandrus Use `if` instead of `else if` since the previous block returns?

Reply:
I prefer this

```js
if (foo) {
  return foo;
} else if (bar) {
  return bar;
}
return baz;
```

to this

```js
if (foo) {
  return foo;
}
if (bar) {
  return bar;
}
return baz;
```

I guess because the `else if` indicates the clauses are related rather than distinct and because it is a little more compact. I guess a third option could be

```js
if (foo) {
  return foo;
} if (bar) {
  return bar;
}
return baz;
```

but that's weird, I think.

Reply:
This seems like a classic case for `switch`, no?

Reply:
We could do a `switch`:

```js
switch (state) {
  case 'disconnected':
    room.participants.forEach(participant => {
      participant._unsubscribeTracks();
    });
    room.emit(state, room, error);
    signaling.removeListener('stateChanged', stateChanged);
    break;
  case 'reconnecting':
    room.emit('reconnecting', new MediaConnectionError());
    break;
  default:
    room.emit('reconnected');
}
```